### PR TITLE
core: Update printer 

### PR DESCRIPTION
**Pull request by ShaolunWang**
- Adding `FloatAttr` printing, so that we use `1.000000e+00` instead of `1.0`
- Updating respective tests

### CODE REVIEW

File: xdsl/printer.py [hunk 1]
```diff
@@ -208,12 +209,12 @@ def _print_result_value(self, op: Operation, idx: int) -> None:
         self.print("%")
         if val in self._ssa_values:
             name = self._ssa_values[val]
-        elif val.name_hint:
-            curr_ind = self._ssa_names.get(val.name_hint, 0)
+        elif val.name:
```
Comment (superlopuh) [inline xdsl/printer.py:212]: undo this change, please

File: xdsl/printer.py [hunk 2]
```diff
@@ -702,3 +705,20 @@ def print_op(self, op: Operation) -> None:
             op.print(self)
         else:
             self.print_op_with_default_format(op)
+
+    def print_optional_attr(
```
Comment (superlopuh) [inline xdsl/printer.py:709]: can you make separate pull requests for the change here and the one above? Depending on the one you keep in this PR, can you please rename the PR into something more descriptive?

**Review by superlopuh — CHANGES_REQUESTED** [review]

File: xdsl/printer.py (hunk 2 above)
Comment (ShaolunWang) [reply xdsl/printer.py:709]: Once I finish updating mlir I will make the new pr

Comment (superlopuh) [reply xdsl/printer.py:709]: Can you mark it as draft plz?

**Comment by superlopuh** [comment]
> I'm not sure what's the best way to change the ipynb files, should I revert the changes that's automatically added by running the cells in vscode?

The way to do it is by 
1) restarting the kernel so the cell execution indexing starts from 0
2) executing all the cells in the notebook from start to finish, there should be no diff in the execution index
3) removing the weird diff at the end of the ipynb file, the one that replaces just "python" by a bunch of stuff in the metadata

**Comment by superlopuh** [comment]
Looks good! Just needs a rebase, and I think it can be opened for review

**Comment by tobiasgrosser** [comment]
Nice! Yes, it seems ready to open this for review!

**Review by superlopuh — APPROVED** [review]

**Review by math-fehr — APPROVED** [review]
:+1:

**Comment by math-fehr** [comment]
Can you just update the notebooks? Then I'll merge it!
